File: internal/store/memory/agents.go
package memory

import (
	"context"
	"slices"
	"time"

	"github.com/oklog/ulid/v2"
	"github.com/rakunlabs/at/internal/service"
)

func (m *Memory) ListAgents(_ context.Context) ([]service.Agent, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	result := make([]service.Agent, 0, len(m.agents))
	for _, a := range m.agents {
		result = append(result, a)
	}

	slices.SortFunc(result, func(a, b service.Agent) int {
		if a.Name < b.Name {
			return -1
		}
		if a.Name > b.Name {
			return 1
		}
		return 0
	})

	return result, nil
}

func (m *Memory) GetAgent(_ context.Context, id string) (*service.Agent, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	a, ok := m.agents[id]
	if !ok {
		return nil, nil
	}

	return &a, nil
}

func (m *Memory) CreateAgent(_ context.Context, agent service.Agent) (*service.Agent, error) {
	id := ulid.Make().String()
	now := time.Now().UTC().Format(time.RFC3339)

	rec := service.Agent{
		ID:            id,
		Name:          agent.Name,
		Description:   agent.Description,
		Provider:      agent.Provider,
		Model:         agent.Model,
		SystemPrompt:  agent.SystemPrompt,
		Skills:        agent.Skills,
		MCPs:          agent.MCPs,
		MaxIterations: agent.MaxIterations,
		ToolTimeout:   agent.ToolTimeout,
		CreatedAt:     now,
		UpdatedAt:     now,
		CreatedBy:     agent.CreatedBy,
		UpdatedBy:     agent.UpdatedBy,
	}

	m.mu.Lock()
	m.agents[id] = rec
	m.mu.Unlock()

	return &rec, nil
}

func (m *Memory) UpdateAgent(_ context.Context, id string, agent service.Agent) (*service.Agent, error) {
	now := time.Now().UTC().Format(time.RFC3339)

	m.mu.Lock()
	defer m.mu.Unlock()

	existing, ok := m.agents[id]
	if !ok {
		return nil, nil
	}

	existing.Name = agent.Name
	existing.Description = agent.Description
	existing.Provider = agent.Provider
	existing.Model = agent.Model
	existing.SystemPrompt = agent.SystemPrompt
	existing.Skills = agent.Skills
	existing.MCPs = agent.MCPs
	existing.MaxIterations = agent.MaxIterations
	existing.ToolTimeout = agent.ToolTimeout
	existing.UpdatedAt = now
	existing.UpdatedBy = agent.UpdatedBy

	m.agents[id] = existing

	return &existing, nil
}

func (m *Memory) DeleteAgent(_ context.Context, id string) error {
	m.mu.Lock()
	delete(m.agents, id)
	m.mu.Unlock()

	return nil
}

File: _ui/src/lib/api/agents.ts
import axios from 'axios';
import type { ListResult, ListParams } from './types';

const api = axios.create({ baseURL: 'api/v1' });

export interface AgentConfig {
  description: string;
  provider: string;
  model: string;
  system_prompt: string;
  skills: string[];
  mcp_sets: string[];
  mcp_urls: string[];
  builtin_tools: string[];
  max_iterations: number;
  tool_timeout: number;
  confirmation_required_tools?: string[];
  avatar_seed?: string;
}

export interface Agent {
  id: string;
  name: string;
  config: AgentConfig;
  created_at: string;
  updated_at: string;
  created_by: string;
  updated_by: string;
}

export async function listAgents(params?: ListParams): Promise<ListResult<Agent>> {
  const res = await api.get<ListResult<Agent>>('/agents', { params });
  return res.data;
}

export async function getAgent(id: string): Promise<Agent> {
  const res = await api.get<Agent>(`/agents/${id}`);
  return res.data;
}

export async function createAgent(data: Partial<Agent>): Promise<Agent> {
  const res = await api.post<Agent>('/agents', data);
  return res.data;
}

export async function updateAgent(id: string, data: Partial<Agent>): Promise<Agent> {
  const res = await api.put<Agent>(`/agents/${id}`, data);
  return res.data;
}

export async function deleteAgent(id: string): Promise<void> {
  await api.delete(`/agents/${id}`);
}

// ─── Import / Export ───

export async function exportAgent(id: string): Promise<string> {
  const res = await api.get<string>(`/agents/${id}/export`, { responseType: 'text' as any });
  return res.data;
}

export async function exportAgentJSON(id: string): Promise<Partial<Agent>> {
  const res = await api.get<Partial<Agent>>(`/agents/${id}/export-json`);
  return res.data;
}

export async function importAgent(markdownContent: string): Promise<Agent> {
  const res = await api.post<Agent>('/agents/import', markdownContent, {
    headers: { 'Content-Type': 'text/markdown' },
  });
  return res.data;
}

export async function previewImportAgent(markdownContent: string): Promise<Partial<Agent>> {
  const res = await api.post<Partial<Agent>>('/agents/import/preview', markdownContent, {
    headers: { 'Content-Type': 'text/markdown' },
  });
  return res.data;
}
